### PR TITLE
Simplify result_of and friends

### PR DESCRIPTION
**Pull request by K-ballo**
Rework `result_of`, `is_callable`, `invoke` implementations for readability.

### CODE REVIEW

File: hpx/lcos/local/dataflow.hpp [hunk 1]
```diff
@@ -81,13 +81,13 @@ namespace hpx { namespace lcos { namespace local
         template <typename Policy, typename Func, typename Futures>
         struct dataflow_frame //-V690
           : hpx::lcos::detail::future_data<
-                typename hpx::util::invoke_fused_result_of<
+                typename hpx::util::detail::fused_result_of<
```
Comment (sithhell) [inline hpx/lcos/local/dataflow.hpp]: Why did fused_result_of move into the detail namespace while invoke_fused still resides in hpx::util?

Comment (K-ballo) [inline hpx/lcos/local/dataflow.hpp]: Because `decltype` support is required now, result traits and the TR1 protocol become obsolete.

Comment (sithhell) [inline hpx/lcos/local/dataflow.hpp]: Ahh, makes sense.

**Comment by sithhell** [comment]
LGTM.

**Pull request merged**